Spec/RealtimeClient.swift

```diff
@@ -352,6 +352,478 @@ class RealtimeClient: QuickSpec {
+                            expect(connectedMessages).to(haveCount(2))
+
+                            guard let connectedAfterAuth = connectedMessages.last, connectionDetailsAfterAuth = connectedAfterAuth.connectionDetails else {
+                                fail("Missing CONNECTED protocol message after AUTH protocol message"); partialDone(); return
+                            }
+
+                            expect(client.auth.clientId).to(beNil())
+                            expect(connectionDetailsAfterAuth.clientId).to(beNil())
+                            expect(client.connection.key).to(equal(connectionDetailsAfterAuth.connectionKey))
+                            partialDone()
+                        }
+
+                        client.auth.authorize(nil, options: nil) { tokenDetails, error in
+                            expect(error).to(beNil())
+                            guard let tokenDetails = tokenDetails else {
+                                fail("TokenDetails is nil"); partialDone(); return
+                            }
+                            expect(tokenDetails.token).toNot(equal(testToken))
+                            partialDone()
+                        }
+                    }
+
+                    expect(client.auth.tokenDetails?.token).toNot(equal(testToken))
+                }
+
+                // RTC8a1 - part 2
+                it("performs an upgrade of capabilities without any loss of continuity or connectivity during the upgrade process") {
+                    let options = AblyTests.commonAppSetup()
+                    options.autoConnect = false
+                    let testToken = getTestToken(capability: "{\"test\":[\"subscribe\"]}")
+                    options.token = testToken
+                    let client = ARTRealtime(options: options)
+                    defer { client.dispose(); client.close() }
+                    client.setTransportClass(TestProxyTransport.self)
+
+                    waitUntil(timeout: testTimeout) { done in
+                        client.connection.once(.Connected) { stateChange in
+                            expect(stateChange?.reason).to(beNil())
+                            done()
+                        }
+                        client.connect()
+                    }
+
+                    waitUntil(timeout: testTimeout) { done in
+                        let partialDone = AblyTests.splitDone(2, done: done)
+
+                        client.connection.once(.Connected) { stateChange in
+                            guard let stateChange = stateChange else {
+                                fail("ConnectionStateChange is nil"); partialDone(); return
+                            }
+                            expect(stateChange.previous).to(equal(ARTRealtimeConnectionState.Connected))
+                            expect(stateChange.reason).to(beNil())
```

Comment:
I don't think you're checking sufficiently that the upgrade has occurred.  I would attempt to attach to channel `foo` with the original token, that should fail because of lack of permissions, then upgrade the token and clientId (we allow upgrade from anonymous to identified), then attempt attach to `foo` which should now work and also check the CONNECTED ConnectionDetails clientId is set correctly.